### PR TITLE
rec: Move to several distinct Bogus states, for easier debugging

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This is especially useful after the fact, when the Bogus state is retrieved from a cache and we don't have any clue left as to how we ended up with that state.
Documentation is missing, and I wouldn't be surprised to see some regression tests fail.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

**Comment by omoerbeek** [comment]
Love this. Puzzled by the sudden Travis failure....

**Comment by rgacogne** [comment]
My money is on a dependency error that has already been fixed for the travis failure:
```
ERROR: Package 'PyHamcrest' requires a different Python: 2.7.14 not in '>=3.5'
```
I restarted it.

**Review by omoerbeek — APPROVED** [review]

**Comment by rgacogne** [comment]
I wonder if it would make sense to have a look at the cases covered in rfc8914 before merging this? Perhaps we could add some Bogus states that would make the implementation of rfc8914 easier.

**Comment by rgacogne** [comment]
I added more states to better match the extended error codes defined in rfc8914.

**Comment by rgacogne** [comment]
Otto, are you OK with me merging this?

File: pdns/validate.cc [hunk 1]
```diff
@@ -952,6 +982,64 @@ void validateDNSKeysAgainstDS(time_t now, const DNSName& zone, const dsmap_t& ds
       //        if(validkeys.empty()) cerr<<"did not manage to validate DNSKEY set based on DS-validated KSK, only passing KSK on"<<endl;
     }
   }
+
+  if (validkeys.size() < tkeys.size()) {
+    /* so we failed to validate the whole set, let's try to find out why exactly */
```
Comment (pieterlexis) [inline pdns/validate.cc:987]: What if one key is bad but the rest are good? do we still bail with a Bogus?

Comment (rgacogne) [reply pdns/validate.cc:987]: You mean if we did not find at least one DNSKEY, corresponding to a DS, that validates the signature of the whole DNSKEY RRset? Yes

**Comment by omoerbeek** [comment]
> Otto, are you OK with me merging this?

yes

**Pull request merged**